### PR TITLE
Use point cloud wraper in ring outlier filter

### PR DESCRIPTION
**Pull request by taichiH**
## Description
To remove pcl dependency change point cloud interface to point_cloud_msg_wrapper in ring outlier filter.

This PR depend on https://github.com/tier4/AutowareArchitectureProposal.iv/pull/784.

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [pull request guidelines][pull-request-guidelines]
- [x] Code follows [coding guidelines][coding-guidelines]
- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code follows [coding guidelines][coding-guidelines]
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets
- [ ] Write [release notes][release-notes]

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[clang-tidy-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/ClangTidyGuideline/
[coding-guidelines]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/1194394777/T4
[pull-request-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/PullRequestGuideline/
[release-notes]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/563774416
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by taichiH** [comment]
I calculated the processing time of the filter function in the same rosbag scene using std::chrono.
- before merge this: 17ms
- after merge this: 8ms

File: common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp
Comment (kenji-miyake) [inline common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp:34]: We should consider the rule of C++ namespace. I think it's good to make it the same as the package name.

File: common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp
Comment (kenji-miyake) [inline common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp:22]: I guess this header guard's name is wrong. But it's not a big problem since it can be automatically fixed later.

**Comment by kenji-miyake** [comment]
You should add tests if you'd like to add these files into `autoware_utils`.
As there are a lot of things to be discussed, I'll set this PR as a draft.

File: common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp
Comment (taichiH) [reply common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp:34]: This file is copied from Autoware.Auto so I don't want to change the code if possible.

https://gitlab.com/autowarefoundation/autoware.auto/AutowareAuto/blob/master/src/common/autoware_auto_common/include/common/types.hpp

Comment (kenji-miyake) [reply common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp:34]: Then, unfortunately, I think you can't merge this to `autoware_utils` because it's not consistent.

**Comment by taichiH** [comment]
@kenji-miyake I'll create separate package in `common/point_types`. Is it Okay?

**Comment by kenji-miyake** [comment]
@taichiH Mm, we should also discuss the license. Let's consult with @yukkysaito as well.

**Comment by taichiH** [comment]
@kenji-miyake I divied pull request to  https://github.com/tier4/AutowareArchitectureProposal.iv/pull/784. Let's dicuss about point_types on another PR.

File: common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp
Comment (taichiH) [reply common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp]: It was resolved by https://github.com/tier4/AutowareArchitectureProposal.iv/pull/784

File: common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp
Comment (taichiH) [reply common/util/autoware_utils/include/autoware_utils/point_types/float_comparisons.hpp]: It was resolved by https://github.com/tier4/AutowareArchitectureProposal.iv/pull/784

**Comment by kenji-miyake** [comment]
LGTM, but I think I don't have the right to approve this. If somebody in the sensing team approves, I guess it's okay.

**Comment by taichiH** [comment]
red points: input pointcloud
green points: ring outlier filtered pointcloud

Screenshot from 2021-12-17 17-31-09

**Comment by taichiH** [comment]
Sorry, I change pr to draft because I found a mistake.

**Comment by taichiH** [comment]
I fixed mistakes in https://github.com/tier4/AutowareArchitectureProposal.iv/pull/780/commits/f09333fbd2e099868dbc80fafae6386289f9dc80, https://github.com/tier4/AutowareArchitectureProposal.iv/pull/780/commits/1b4563d98a1154dd07b66f3fc51c94bd27a52d0d

**Review by yukkysaito — APPROVED** [review]

**Comment by yukkysaito** [comment]
> red points: input pointcloud green points: ring outlier filtered pointcloud
> 
> Screenshot from 2021-12-17 17-31-09

@taichiH did you mistake?

yellow : input pointcloud 
red: ring outlier filtered pointcloud

**Comment by yukkysaito** [comment]
@taichiH Please give us the results of CPU load before and after the change.

**Comment by taichiH** [comment]
@yukkysaito Oh, sorry! you are right.

**Comment by taichiH** [comment]
@yukkysaito 
Here is a result. I measure loads and stores by vtune but I don't seem to be a meaningful value because the variation is large every time it is measured...

These are results when run autoware and measure stores and loads.

before
Screenshot from 2021-12-24 00-23-16

after
Screenshot from 2021-12-24 00-19-35

**Comment by yukkysaito** [comment]
Great. 
- The memory bound is decreasing, and the dominance of memory accesses in the overall process is decreasing.
- The number of STORES is decreasing.
- But, cash mistakes are on the rise.

@taichiH If you have time, could you please take a look at the execution time for one callback?

**Pull request merged**
